CI: Added the automatic labeler for the PR opened

Ronitsabhaya75 opened the pull request:
- Added the labeler pipeline for PR to understand what particular file is changed.

Commit:
update the label check for the automatic perfoming

Commit:
updated the white space

Commit:
updated the labeler

Commit:
fix: remove corrupted text from labeler.yml and fix whitespace issues

Commit:
updated the labeler yml

LilithHafner reviewed (commented):
Thanks! This is a great idea and I'd like to try it.

Let's start very conservative
- Only apply labels that already exist
- Only apply labels when it is certain that whenever that file is changed the label is appropriate.

We can always expand the scope of auto-labeling if this proves useful and not annoying/spammy/false positives. Please select 1-3  label application rules to start with and save the rest for later.

Is there any way to test this on a bunch of old PRs to see how well it matches manual categorization?

Thanks for contributing!

DilumAluthge commented:
Running `on: pull_request:` won't work for PRs from forks, because the CI job won't have any write access.

If we run `on: pull_request_target:`, the CI job will have write access, but in such a situation, I very much do NOT want to do an `actions/checkout`, because this is an easy way to get pwned: ([source](https://securitylab.github.com/resources/github-actions-preventing-pwn-requests/))

> TL;DR: Combining pull_request_target workflow trigger with an explicit checkout of an untrusted PR is a dangerous practice that may lead to repository compromise.

So, to make life easier for myself as a maintainer, if we do switch to running `on: pull_request_target:`, I don't want us to have any checkouts anywhere in the CI file - this makes it easier to review the CI file for security.

DilumAluthge commented on .github/workflows/labeler.yml:
```diff
@@ -0,0 +1,27 @@
+    runs-on: ubuntu-latest
+    timeout-minutes: 5
+    steps:
+      - name: Checkout the JuliaLang/julia repository
+        uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
+        with:
+          persist-credentials: false
```
https://github.com/JuliaLang/julia/pull/59449#issuecomment-3240482994

```suggestion
```

DilumAluthge commented on .github/workflows/labeler.yml:
```diff
@@ -0,0 +1,27 @@
+  pull-requests: write
+
+on:
+  pull_request:
```
```suggestion
  pull_request_target:
```

Commit:
updated the code with changes requested

Commit:
fix: resolve whitespace issues in labeler configuration

- Remove trailing whitespace from labeler.yml line 18
- Add missing trailing newline to labeler.yml
- Remove trailing whitespace from workflows/labeler.yml line 19

DilumAluthge commented on .github/labeler.yml:
```diff
@@ -0,0 +1,202 @@
+      - 'base/show.jl'
+
+# Testing and CI
+"tests":
+  - changed-files:
+    - any-glob-to-any-file: 'test/**/*'
```
Yeah, I agree - I don't think it will be super useful for every PR that modifies a test to have the `tests` label.

I'd suggest getting rid of this for now - we can always re-evaluate later.

DilumAluthge commented on .github/labeler.yml:
```diff
@@ -0,0 +1,202 @@
+    - any-glob-to-any-file:
+      - '.github/**/*'
+      - '.buildkite/**/*'
+      - 'contrib/**/*'
```
Yeah, agreed - let's remove the `'contrib/**/*'` line.

Commit:
refactor: address DilumAluthge feedback - reduce labeler scope

Ronitsabhaya75 commented:
@DilumAluthge I refactored the code changed it to specific things you need

Ronitsabhaya75 commented on .github/labeler.yml:
```diff
@@ -0,0 +1,16 @@
+# Configuration for automatic PR labeling based on file paths
+# See: https://github.com/actions/labeler
+comm#
```
yes it was typo my bad havent gone through this

Commit:
fix: resolve typo and add permission comment

Ronitsabhaya75 commented:
@DilumAluthge @LilithHafner can you please review this updated changes

DilumAluthge commented on .github/workflows/labeler.yml:
```diff
@@ -0,0 +1,31 @@
+
+on:
+  pull_request_target:
+    types: [opened, synchronize]
```
@Ronitsabhaya75 Can you change this to just `opened`? Otherwise, the action will run every time someone pushes to a PR, which I think is not what we want.

Commit:
refactor: address all DilumAluthge feedback in single PR

- Change labeler trigger to only 'opened' (not synchronize) to avoid re-running on every push
- Keep LabelCheck.yml improvements in this PR as requested
- Add ready_for_review trigger to LabelCheck for better workflow
- All security concerns addressed: pull_request_target + no checkout

Commit:
refactor: address all DilumAluthge feedback in single PR

DilumAluthge commented:
@Ronitsabhaya75 For the `.github/workflows/LabelCheck.yml` file, can you keep the change to line 1 (changing `name: Labels` to `name: Label Check`), but remove all other changes? You can put those changes in a separate PR, if you want.

Commit:
fix: minimal LabelCheck.yml change per DilumAluthge feedback

- Keep only the name change from 'Labels' to 'Label Check' as requested
- Remove all other modifications (draft condition, ready_for_review trigger)
- Restore original version comment and formatting
- Other LabelCheck improvements can be addressed in separate PR if needed

Ronitsabhaya75 commented:
> @Ronitsabhaya75 For the `.github/workflows/LabelCheck.yml` file, can you keep the change to line 1 (changing `name: Labels` to `name: Label Check`), but remove all other changes? You can put those changes in a separate PR, if you want.

I did the changes you asked

Commit:
Rename workflow from 'Label Check' to 'Labels'

DilumAluthge commented on .github/workflows/labeler.yml:
```diff
@@ -0,0 +1,27 @@
+        uses: actions/labeler@8558fd74291d67161a8a78ce36a881fa63b766a9 # v5.0.0
+        with:
+          repo-token: "${{ secrets.GITHUB_TOKEN }}"
+          configuration-path: ".github/labeler.yml"
```
Looks like this has been addressed now.

DilumAluthge commented:
I think it would be nice to test this out before merging. Unfortunately, because we use `pull_request_target`, this action won't run until we merge.

@Ronitsabhaya75 Perhaps you can set up a separate test repo, and put this action in it, and do some test PRs, to make sure it works as expected?

Commit:
Merge branch 'master' into master

Commit:
docs: add educational comment about pull_request_target

- Add LilithHafner's suggested comment explaining pull_request_target behavior
- Clarifies that action runs using master version, not branch version
- Explains trusted execution context and security implications
- Helps future maintainers understand the security considerations

DilumAluthge commented:
Can you link some of the test PRs here?
